### PR TITLE
Feature/user registration

### PR DESCRIPTION
**Pull request by abaniuszewicz**
Just for review purposes.

### CODE REVIEW

File: backend/Modules/Users/HitMeApp.Users/Infrastructure/IUserRepository.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿using System.Threading.Tasks;
+using HitMeApp.Shared.DDD;
+using HitMeApp.Users.Models;
+
+namespace HitMeApp.Users.Infrastructure
+{
+    internal interface IUserRepository : IRepository
+    {
+        public Task Add(User user);
+        public bool Exists(string email);
```
Comment (Zuravvski) [inline backend/Modules/Users/HitMeApp.Users/Infrastructure/IUserRepository.cs]: `bool` -> `Task<bool>`. Unfortunatelly most of the repositories will be called asynchronously, and it kind of enforces the signature to follow the pattern. It's easier to handle asynchronous interfaces synchronously than the other way around.

File: backend/Modules/Users/HitMeApp.UsersTests/Cryptography/CryptoTests.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
+﻿using System;
+using HitMeApp.Users.Cryptography;
+using NUnit.Framework;
+
+namespace HitMeApp.UsersTests.Cryptography
```
Comment (Zuravvski) [inline backend/Modules/Users/HitMeApp.UsersTests/Cryptography/CryptoTests.cs]: To follow the general convention the project should've been called: `HitMeApp.Users.Tests`

File: backend/Modules/Users/HitMeApp.Users/Exceptions/PasswordNotValidException.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
+﻿using System;
+using System.Runtime.Serialization;
+
+namespace HitMeApp.Users.Exceptions
+{
+    internal class PasswordNotValidException : Exception
```
Comment (Zuravvski) [inline backend/Modules/Users/HitMeApp.Users/Exceptions/PasswordNotValidException.cs]: I think `InvalidPasswordException` feels like a better name, but treat it optionally :)

File: backend/Modules/Users/HitMeApp.Users/Exceptions/EmailNotValidException.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
+﻿using System;
+using System.Runtime.Serialization;
+
+namespace HitMeApp.Users.Exceptions
+{
+    internal class EmailNotValidException : Exception
```
Comment (Zuravvski) [inline backend/Modules/Users/HitMeApp.Users/Exceptions/EmailNotValidException.cs]: I think `InvalidEmailException` feels like a better name, but treat it optionally :)

File: backend/Modules/Users/HitMeApp.Users/Cryptography/Crypto.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using System.Security.Cryptography;
+
+namespace HitMeApp.Users.Cryptography
+{
+    internal static class Crypto
```
Comment (Zuravvski) [inline backend/Modules/Users/HitMeApp.Users/Cryptography/Crypto.cs]: 1. Keeping it static makes no sense in the long run as it would only be needed to be instantiated when the user wants to login.
2. It makes sense to wrap whatever the whatever encryption method we use in our own adapter to provide a convenient way to replace it. It's not that trivial for the app that is developed for the long time, but in the beginning phase of the project it's extremely convient to have it hooked up in a flexible manner.
3. .NET finally provided some convenient implementation of its own called `IPasswordHasher<T>`. I think we should wrap it around our own interface. 

To sum it up: I would wrap `IPasswordHasher<T>` in our own adapter and register this adapter as `InstancePerLifetimeScope` in module's IoC container.

File: backend/Modules/Users/HitMeApp.Users/Exceptions/UserAlreadyExistsException.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
+﻿using System;
+using System.Runtime.Serialization;
+
+namespace HitMeApp.Users.Exceptions
+{
+    internal class UserAlreadyExistsException : Exception
```
Comment (Zuravvski) [inline backend/Modules/Users/HitMeApp.Users/Exceptions/UserAlreadyExistsException.cs]: All infrastructure related exceptions should inherit `ApplicationException` provided in https://github.com/Somsiady-Inc/HitMeApp/pull/15 when it's merged. 

That would not only make exceptions more consistent, but will be really important for some coding magic later on.

**Review by Zuravvski — CHANGES_REQUESTED** [review]

File: backend/Modules/Users/HitMeApp.UsersTests/Cryptography/CryptoTests.cs (hunk 2 above)
Comment (abaniuszewicz) [reply backend/Modules/Users/HitMeApp.UsersTests/Cryptography/CryptoTests.cs]: Done

File: backend/Modules/Users/HitMeApp.Users/Exceptions/PasswordNotValidException.cs (hunk 3 above)
Comment (abaniuszewicz) [reply backend/Modules/Users/HitMeApp.Users/Exceptions/PasswordNotValidException.cs]: Done

File: backend/Modules/Users/HitMeApp.Users/Exceptions/EmailNotValidException.cs (hunk 4 above)
Comment (abaniuszewicz) [reply backend/Modules/Users/HitMeApp.Users/Exceptions/EmailNotValidException.cs]: Done

File: backend/Modules/Users/HitMeApp.Users/Infrastructure/InMemoryUserRepository.cs [hunk 7]
```diff
@@ -15,9 +15,9 @@ public Task Add(User user)
             return Task.CompletedTask;
         }
 
-        public bool Exists(string email)
+        public Task<bool> Exists(string email)
         {
-            return _users.Any(u => u.Email == email);
+            return new Task<bool>(() => _users.Any(u => u.Email == email));
```
Comment (Zuravvski) [inline backend/Modules/Users/HitMeApp.Users/Infrastructure/InMemoryUserRepository.cs]: `Task.FromResult()`

File: backend/Modules/Users/HitMeApp.Users/Cryptography/Crypto.cs (hunk 5 above)
Comment (abaniuszewicz) [reply backend/Modules/Users/HitMeApp.Users/Cryptography/Crypto.cs]: Not applicable anymore. As suggested, I've used default `IPasswordHasher<T>` implementation and got rid of our custom `Crypto` class. No need to reinvent the wheel and overcomplicate things at the very beginning. Default implementation has been registered in IoC container.

File: backend/Modules/Users/HitMeApp.Users/Exceptions/UserAlreadyExistsException.cs (hunk 6 above)
Comment (abaniuszewicz) [reply backend/Modules/Users/HitMeApp.Users/Exceptions/UserAlreadyExistsException.cs]: Done

**Review by Zuravvski — APPROVED** [review]
